Accurate lead target

KyleAnthonyShepherd opened the pull request:
PR to address this issue:
https://github.com/beyond-all-reason/spring/issues/746

Seeking advice on how this change should be structured.
Currently, this PR just applies to gravity affected `WEAPON_EXPLOSIVE_PROJECTILE`, but issues 746 applies to all weapons with a non-zero travel time. And only addresses low trajectory units. 

This PR adds a new weapondef boolean `accurateLeading`, to use fixed point iterations to calculate accurate intercept positions.  
The current method is inaccurate for targets moving towards or away from the unit, because the current method does not take into account changes in projectile travel time as the target moves closer or further away. 
The gameplay implication is units with slow projectiles, such as the BAR fatboy, leads approaching fast units too much, frequently firing too far ahead and friendly firing. 

Fundamentally, this target leading problem for gravity-affected projectiles involves solving for the smallest positive zero of a quartic polynomial, and I found fixed point iterations to be the most stable way of finding this value, or exiting if no positive real solution exists.

sprunk commented on rts/Sim/Weapons/Weapon.cpp:
```diff
@@ -1268,12 +1271,57 @@ float3 CWeapon::GetUnitLeadTargetPos(const CUnit* unit) const
+				float temp1 = 1.0f;
+				float temp2 = 1.0f;
+				float cc = 1.0f;
+				for (int ii = 0; ii < 32; ii++) {
```
Perhaps `accurateLeading` should be a number, and then
```suggestion
				for (int ii = 0; ii < accurateLeading; ii++) {
```

lhog replied:
A future generation of readers will appreciate if you could outline the math details of iteration method you've implemented or you can leave a link to some reliable source, like wikipedia.

KyleAnthonyShepherd replied:
Reviewed and further elaborated on the iteration method

sprunk commented on rts/Sim/Weapons/Weapon.cpp:
```diff
@@ -1268,12 +1271,57 @@ float3 CWeapon::GetUnitLeadTargetPos(const CUnit* unit) const
+	float predictTime = GetPredictedImpactTime(unit->pos);
 	const float predictMult = mix(predictSpeedMod, 1.0f, weaponDef->predictBoost);
 
+	if (accurateLeading == true) {
```
Consider extracting this chunk to a function that returns the new dist and predictTime

blue-bear-tech commented on rts/Sim/Weapons/Weapon.cpp:
```diff
@@ -1268,13 +1271,159 @@ float3 CWeapon::GetUnitLeadTargetPos(const CUnit* unit) const
+			// cannon has no gravity
+			// non-parabolic projectiles are much easier
+			// simple quadratic equation
+			float3 unitSpeed = unit->speed * predictMult;
+			float3 dist = unit->pos - weaponMuzzlePos;
+			float aa = unitSpeed.dot(unitSpeed) - (weaponDef->projectilespeed) * (weaponDef->projectilespeed);
+			float bb = 2 * (dist.dot(unitSpeed));
+			float cc = dist.dot(dist);
+			float temp1 = 4 * aa * cc;
+			float temp2 = (bb * bb);
+			if (temp1 >= temp2) {
+				// this triggers if the target cannot be intercepted
+				// units can get out of range before the slow projectile can hit it
+				return predictTime;
+			}
+			predictTime = (-bb - math::sqrt(temp2 - temp1)) / (2 * aa);
```
My C++ is a bit rusty but won't `predictTime` hit infinity at some point if `aa` is really small?

KyleAnthonyShepherd replied:
Thanks for the comments. :)
I spun this out into a dedicated function `GetSafeInterceptTime` to try to catch numerically instabilities like the ones you pointed out, and more.

sprunk commented on rts/Sim/Weapons/Weapon.cpp:
```diff
@@ -1275,6 +1275,69 @@ float3 CWeapon::GetUnitLeadTargetPos(const CUnit* unit) const
+		else {
+			// check catastrophic case of aa=0 and bb=0 
+			// target speed equal to projectile speed, and target is moving tangentally
+			if ((std::abs(aa) < (1.0 / (2 ^ 16))) && (std::abs(bb) < (1.0 / (2 ^ 16)))) {
```
2 ^ 16 is 18

sprunk commented on rts/Sim/Weapons/Weapon.cpp:
```diff
@@ -1275,6 +1275,69 @@ float3 CWeapon::GetUnitLeadTargetPos(const CUnit* unit) const
+	// while also being numerically stable
+	// We also know c is strictly positive.
+
+	// case 1, aa <0, target speed is less than projectile speed
+	// guaranteed existence of a positive solution
+	if (aa < -1) {
```
comment says < 0 but logic says < -1

sprunk commented on rts/Sim/Weapons/Weapon.cpp:
```diff
@@ -1370,6 +1433,8 @@ float CWeapon::GetAccuratePredictedImpactTime(const CUnit* unit) const
 			// non-parabolic projectiles are much easier
 			// simple quadratic equation
+			predictTime = GetSafeInterceptTime(unit, predictMult);
+			/*
```
don't leave commented out code, or at least leave a hint on how it may become useful in the future

KyleAnthonyShepherd commented:
Profiling with Tracy now.
CWeapon::GetLeadVec
Base game (UnitWeapondef boolean `accurateLeading = false`) = 100 ns mean
`GetSafeInterceptTime` branch (non-gravity affected projectile) = 350 ns mean
`Iterative Solution` branch (gravity affected projectile) = 500 ns mean

Compared to `CCobInstance::AimWeapon` taking an average of 16000 ns, target leading is cheap. 
Games enabling the `accurateLeading` UnitWeapondef boolean on all of their units should be a minimal overall performance impact, but keeping it default off preserves backwards compatibility and highest performance. 
![image](https://github.com/beyond-all-reason/spring/assets/44480662/178b2014-ab29-4d2e-b7ea-c9044f8ec94c)

I will continue addressing comments and cleaning up comments.

sprunk commented on rts/Sim/Weapons/Weapon.cpp:
```diff
@@ -1297,42 +1297,46 @@ float CWeapon::GetSafeInterceptTime(const CUnit* unit, float predictMult) const
 		// target speed equal to projectile speed, and target is moving tangentally
-		if ((std::abs(aa) < (1.0 / (2 ^ 16))) && (std::abs(bb) < (1.0 / (2 ^ 16)))) {
-			return 0.0;
+		if ((std::abs(aa) < (std::pow(2, -16))) && (std::abs(bb) < (std::pow(2, -16)))) {
```
consider making sure this is a compile-time constant, for example `constexpr auto VELOCITY_EPS = std::pow(2, -16);`

FortK commented on rts/Sim/Weapons/Weapon.cpp:
```diff
@@ -1275,13 +1276,241 @@ float3 CWeapon::GetUnitLeadTargetPos(const CUnit* unit) const
+		// standard quadratic formula
+		predictTime = (-bb - math::sqrt(temp2 - temp1)) / (2 * aa);
+	}
+	// case 1b, aa is a small value, "inverted" standard quadratic formula works better, and extra check needed
+	else if (aa <= 0) {
+		// check catastrophic case of aa=0 and bb=0 
+		// target speed equal to projectile speed, and target is moving tangentally
+		if ((std::abs(aa) < (std::pow(2, -16))) && (std::abs(bb) < (std::pow(2, -16)))) {
+			return -1.0;
+		}
+		// use Citardauq Formula if aa is small
+		predictTime = (2 * cc) / (-bb + math::sqrt(temp2 - temp1));
+	}
```
What about the no-solution condition of `aa=0 and bb>0`? This makes the denominator `-bb + sqrt( bb*bb - 4*aa*cc ) = - bb + sqrt( bb^2 - 0 ) = - bb + bb = 0`, which means we divide by zero and get Infinity. It doesn't happen with `aa=0 and bb<0` as it reduces to `- (- bb) + sqrt ( (- bb)^2 ) = bb + sqrt ( bb^2 ) = bb + bb = 2 * bb`

KyleAnthonyShepherd commented:
I think this is ready for another review.
Addressed more comments. Added more code comments. 
Also using `weaponDef->gravityAffected` instead of just checking for `cannon` weapontype. 
Ran a test in BAR, giving every unit and their weapons the new `accurateLeading` tag, arm vs cor, and everything seemed to work for me.

lhog commented:
Looks like a useful change, let's reactivate the work here.

KyleAnthonyShepherd commented:
https://github.com/user-attachments/assets/cc95593d-b662-4263-85d9-8c5b6bd044e5

Reviewed the math, checked for divide by zero cases. Changed `accurateLeading` to an int as suggested.

Also, as seen in the .mp4, accurateLeading now also applies to high trajectory shots.

KyleAnthonyShepherd commented on rts/Sim/Weapons/Weapon.cpp:
```diff
@@ -33,6 +34,8 @@
 
 #include "System/Misc/TracyDefs.h"
 
+constexpr float SAFE_INTERCEPT_EPS = (1.0 / 65536);
```
65536 is 2^16
1/65536 = 1.52x10^-5

Largely arbitrary. 

cmp_eps() = 1.00x10^-4

And looks like that is good enough for the matrix math in Recoil, so should be a good enough threshold for "hey, this number is getting small, you should probably be careful around divisions"

sprunk commented on rts/Sim/Weapons/Weapon.cpp:
```diff
@@ -1367,14 +1372,230 @@ float3 CWeapon::GetUnitLeadTargetPos(const CUnit* unit) const
+		// weapon not gravity affected
+		// non-parabolic projectiles can be directly calculated
+		// just need to solve the quadratic equation, in a numerically safe way
+		const float interceptTime = GetSafeInterceptTime(unit, predictMult);
+		if (interceptTime > 0) {
+			predictTime = interceptTime;
+		}
```
looks duplicated with the chunk above

KyleAnthonyShepherd commented on rts/Sim/Weapons/Weapon.cpp:
```diff
@@ -1367,14 +1372,230 @@ float3 CWeapon::GetUnitLeadTargetPos(const CUnit* unit) const
+	float3 dist = unit->pos - weaponMuzzlePos;
+	float aa = unitSpeed.dot(unitSpeed) - (weaponDef->projectilespeed) * (weaponDef->projectilespeed);
+	float bb = 2 * (dist.dot(unitSpeed));
+	float cc = dist.dot(dist);
```
I'll leave as-is.

KyleAnthonyShepherd commented on rts/Sim/Units/UnitDef.h:
```diff
@@ -45,6 +45,7 @@ struct UnitDefWeapon {
 
 	bool fastAutoRetargeting = false; ///< pick new targets as soon as possible, don't wait for slow update
 	bool fastQueryPointUpdate = false;	///< check in with unitscript to get most current query piece before every friendly fire check, don't wait for slow update
+	unsigned int accurateLeading = false;	///< Accurately lead moving targets. 0 = undershoot approaching or retreating targets (default), 1 = exact solution for non-parabolic shots and 1 accuracy iteration for parabolic shots, 2 = one frame accuracy for almost all shots, 2+ = more iterations for more accuracy if needed.
```
Suggestion noted and incorporated.

KyleAnthonyShepherd commented:
@sprunk I plan to merge this in Monday morning unless you have any objections.

sprunk reviewed (approved):
Looks good from a design PoV but I haven't reviewed the math beyond a basic glance.
